samples: cellular: add nrf_cloud_coap_cell_location sample (#23959)
nrfconnect/sdk-nrf: samples/cellular/nrf_cloud_coap_cell_location/README.rst

You must onboard your device to nRF Cloud for this sample to function.
You only need to do this once for each device.
To onboard your device, install `nRF Cloud Utils`_ and follow the instructions in the README.

Configuration
*************
|config|

Configuration options
=====================

Check and configure the following Kconfig options for the sample:

.. _CONFIG_COAP_CELL_CHANGE_CONFIG:

CONFIG_COAP_CELL_CHANGE_CONFIG - Enable changing request configuration
   Set this to use the next combination of ``hi_conf`` and ``fallback`` flags after performing both single- and multi-cell location requests.

.. _CONFIG_COAP_CELL_DEFAULT_DOREPLY_VAL:

CONFIG_COAP_CELL_DEFAULT_DOREPLY_VAL - Enable return of location from cloud
   If enabled, request the cloud to return the location information.

.. _CONFIG_COAP_CELL_DEFAULT_FALLBACK_VAL:

CONFIG_COAP_CELL_DEFAULT_FALLBACK_VAL - Enable fallback to coarse location
   If enabled and the location of the cell tower or Wi-Fi® access points cannot be found, return area-level location based on the cellular tracking area code.
   Otherwise an error will be returned indicating location is not known.

.. _CONFIG_COAP_CELL_DEFAULT_HICONF_VAL:

> CONFIG_COAP_CELL_DEFAULT_HICONF_VAL - Enable high confidence result

Reviewer: ```suggestion
CONFIG_COAP_CELL_DEFAULT_HICONF_VAL - Enable high-confidence result
```

.. _nrf_cloud_coap_cell_location:

Cellular: nRF Cloud CoAP cellular location
##########################################

.. contents::
   :local:
   :depth: 2

This sample demonstrates how to use the `nRF Cloud CoAP API`_ for nRF Cloud's cellular location service on your device.

Requirements
************

The sample supports the following development kits:

.. table-from-sample-yaml::

.. include:: /includes/tfm.txt

The sample requires an `nRF Cloud`_ account.
Your device must be onboarded to nRF Cloud.
If it is not, follow the instructions in `Device on-boarding <nrf_cloud_coap_cell_location_onboarding>`_.

.. note::
   This sample requires modem firmware v1.3.x or later for an nRF9160 DK, or modem firmware v2.x.x for the nRF9161 and nRF9151 DKs.

Overview
********

After the sample initializes and connects to the network, it enters the single-cell mode and sends a single-cell location request to nRF Cloud using network data obtained from the :ref:`modem_info_readme` library.

> To enable the multi-cell mode, press **Button 1**.

Reviewer 2: Let's not mix user calls with the overview description. This should be mentioned in the Testing section, which is mandatory as per [Sample template](https://docs.nordicsemi.com/bundle/ncs-latest/page/nrf/templates/sample_README.html).

In the multi-cell mode, the sample requests for neighbor cell measurement using the :ref:`lte_lc_readme` library.

> If neighbor cell data is measured, the sample sends a multi-cell location request to nRF Cloud.

Reviewer 2: What performs the measurement and returns the data? Can you rephrase to active voice? Thanks.

> After the sample initializes and connects to the network, it enters the single-cell mode and sends a single-cell location request to nRF Cloud using network data obtained from the :ref:`modem_info_readme` library.
> To enable the multi-cell mode, press **Button 1**.
> In the multi-cell mode, the sample requests for neighbor cell measurement using the :ref:`lte_lc_readme` library.
> If neighbor cell data is measured, the sample sends a multi-cell location request to nRF Cloud.
> Otherwise, the request is single-cell.
> In either mode, the sample sends a new location request if a change in cell ID is detected.

Reviewer 2: ```suggestion
The sample can work in two modes: single-cell mode and multi-cell mode.

After the sample initializes and connects to the network, it enters the single-cell mode and sends a single-cell location request to nRF Cloud.
For this purpose, the sample uses network data obtained from the :ref:`modem_info_readme` library.

In the multi-cell mode, the sample requests for neighbor cell measurement using the :ref:`lte_lc_readme` library.
If neighbor cell data is measured, the sample sends a multi-cell location request to nRF Cloud.
Otherwise, the request is single-cell.

In either mode, the sample sends a new location request if a change in cell ID is detected.
```

See the `nRF Cloud Location Services documentation`_ for additional information.

Limitations
***********

The sample requires the network carrier to provide date and time to the modem.

> Without a valid date and time, the modem cannot generate JWTs with an expiration time.

Reviewer 2: Limitations section is not allowed by [Sample template](https://docs.nordicsemi.com/bundle/ncs-latest/page/nrf/templates/sample_README.html). Move to Troubleshooting under Building and running?

Author: I see no mention of "not allowing" to add sections. This provides background information.

Reviewer 2: > This provides background information.
Indeed, and it provides information about an issue that can be found when the sample is running. Please put it in the Troubleshooting section, as per template.

Author: done

User interface
**************

> Pressing **Button 1** toggles between the single-cell and multi-cell mode.

Reviewer 2: Match style and structure of the [sample template](https://docs.nordicsemi.com/bundle/ncs-latest/page/nrf/templates/sample_README.html#user_interface) for this section.

> Set the :ref:`CONFIG_COAP_CELL_CHANGE_CONFIG <CONFIG_COAP_CELL_CHANGE_CONFIG>` Kconfig option value to try all combinations of the :c:struct:`nrf_cloud_location_config` structure values ``hi_conf`` and ``fallback``.

Reviewer 2: This belongs in Configuration (Configuration options).

.. _nrf_cloud_coap_cell_location_onboarding:

Onboarding your device to nRF Cloud
***********************************

You must onboard your device to nRF Cloud for this sample to function.
You only need to do this once for each device.
To onboard your device, install `nRF Cloud Utils`_ and follow the instructions in the README.

Configuration
*************

|config|

> 

Reviewer: ```suggestion
```
Just remove an extra empty line.

CONFIG_COAP_CELL_DEFAULT_DOREPLY_VAL - Enable return of location from cloud
   If enabled, request the cloud to return the location information.

.. _CONFIG_COAP_CELL_DEFAULT_FALLBACK_VAL:

CONFIG_COAP_CELL_DEFAULT_FALLBACK_VAL - Enable fallback to coarse location
   If enabled and the location of the cell tower or Wi-Fi® access points cannot be found, return area-level location based on the cellular tracking area code.
   Otherwise an error will be returned indicating location is not known.

.. _CONFIG_COAP_CELL_DEFAULT_HICONF_VAL:

CONFIG_COAP_CELL_DEFAULT_HICONF_VAL - Enable high-confidence result
   Enable a 95% confidence interval for the location, instead of the default 68%.

.. _CONFIG_COAP_CELL_SEND_DEVICE_STATUS:

CONFIG_COAP_CELL_SEND_DEVICE_STATUS - Send device status
   Send device status to nRF Cloud on initial connection.

.. include:: /libraries/modem/nrf_modem_lib/nrf_modem_lib_trace.rst
   :start-after: modem_lib_sending_traces_UART_start
   :end-before: modem_lib_sending_traces_UART_end

Building and running
********************

.. |sample path| replace:: :file:`samples/cellular/nrf_cloud_coap_cell_location`

.. include:: /includes/build_and_run_ns.txt

> 

Reviewer: Is there a way of testing the sample? Testing is a mandatory section in a sample doc: https://docs.nordicsemi.com/bundle/ncs-latest/page/nrf/templates/sample_README.html

   :start-after: modem_lib_sending_traces_UART_start
   :end-before: modem_lib_sending_traces_UART_end

Building and running
********************

.. |sample path| replace:: :file:`samples/cellular/nrf_cloud_coap_cell_location`

.. include:: /includes/build_and_run_ns.txt

Testing
=======

|test_sample|

#. |connect_kit|
#. |connect_terminal|
#. Reset the development kit.
#. Observe in the terminal window that the application starts.
   This is indicated by output similar to the following (there is also a lot of additional information about the LTE connection):

   .. code-block:: console

      *** Booting nRF Connect SDK v3.0.99-d1d59238619c ***
      *** Using Zephyr OS v4.1.99-82973b4f6d2a ***
      Attempting to boot slot 0.
      Attempting to boot from address 0x8200.
      I: Trying to get Firmware version
      I: Verifying signature against key 0.
      I: Hash: 0x3e...f9
      I: Firmware signature verified.
      Firmware version 2
      �[00:00:00.254,821] <inf> spi_nor: GD25LE255E@0: 32 MiBy flash
      *** Booting nRF Connect SDK v3.0.99-d1d59238619c ***
      *** Using Zephyr OS v4.1.99-82973b4f6d2a ***

>       [00:00:00.307,281] <inf> nrf_cloud_rest_cell_location_sample: nRF Cloud REST Cellular Location Sample, version: 1.0.0

Reviewer 3: I don't think the logging is correct, it is saying is REST
